### PR TITLE
Add `axis_labels` and `units` to layers and transforms 

### PR DESCRIPTION
**Pull request by Czaki**
This PR add units to our transforms and provides an interface to add them in Layers.

As of this PR, the information is not used by any part of the code base. However, it provides a way for plugins to load and store this information, which in future versions of napari will be used:
- when figuring out which axes are in the viewer (and updating the dims model and displayed axis names)
- when broadcasting axes (for example, displaying a max projection time series (tyx) together with its corresponding volume data (tzyx)
- when adding a scale bar
- by plugins making measurements
- ... and so on.

### CODE REVIEW

File: napari/utils/transforms/_tests/test_transforms.py [hunk 1]
```diff
@@ -1,12 +1,18 @@
...
 from scipy.stats import special_ortho_group
 
 from napari.utils.transforms import Affine, CompositeAffine, ScaleTranslate
 
 transform_types = [Affine, CompositeAffine, ScaleTranslate]
 
+affine_type = [Affine, CompositeAffine]
+
+REG = pint.get_application_registry()
+PIXEL = REG.pixel
```
Comment (jni) [inline napari/utils/transforms/_tests/test_transforms.py:14]: Wait is pixel a built-in unit?? Since when??? (I have a memory of having to manually define it...)

File: napari/utils/transforms/_tests/test_transforms.py [hunk 2]
```diff
@@ -366,3 +372,22 @@ def test_affine_rotate_3d():
...
+
+@pytest.mark.parametrize('AffineType', affine_type)
+def test_set_units(AffineType):
+    assert AffineType(ndim=2, units=('cm', 'mm')).units == (REG.cm, REG.mm)
+
+
+@pytest.mark.parametrize('AffineType', affine_type)
+def test_empty_axis_labels(AffineType):
+    assert AffineType().axis_labels == ('axis 1', 'axis 0')
+    assert AffineType(ndim=3).axis_labels == ('axis 2', 'axis 1', 'axis 0')
```
Comment (jni) [inline napari/utils/transforms/_tests/test_transforms.py]: I think these should be [-3, -2, -1]? This would make them compatible with NumPy indexing.

File: napari/utils/transforms/_units.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+@overload
+def get_units_from_name(
+    units: Sequence[Union[str, pint.Unit]],
+) -> tuple[pint.Unit, ...]: ...
+
+
+def get_units_from_name(units: UnitsLike) -> UnitsInfo:
+    """
+    Convert a string or dict of strings to unyt units.
+    """
```
Comment (jni) [inline napari/utils/transforms/_units.py]:
```suggestion
    """Convert a string or dict of strings to pint units."""
```

File: napari/utils/transforms/transforms.py [hunk 4]
```diff
@@ -491,6 +500,38 @@ def ndim(self) -> int:
...
+        """tuple of axis labels for the layer."""
+        return self._axis_labels
+
+    @axis_labels.setter
+    def axis_labels(self, axis_labels: Optional[Sequence[str]]) -> None:
+        if axis_labels is None:
+            axis_labels = tuple(f'axis {i}' for i in range(self.ndim))[::-1]
+        if len(axis_labels) != self.ndim:
+            raise ValueError(
+                f'{axis_labels=} need to have length ({self.ndim}).'
```
Comment (jni) [inline napari/utils/transforms/transforms.py]:
```suggestion
                f'{axis_labels=} must have length ndim={self.ndim}.'
```

File: napari/utils/transforms/transforms.py [hunk 5]
```diff
@@ -499,6 +540,11 @@ def scale(self) -> npt.NDArray:
         self._setup_decompose_linear_matrix_cache()
         return self._cache_dict['decompose_linear_matrix'][1]
 
+    @property
+    def voxel_size(self) -> tuple[pint.Quantity, ...]:
+        """Return the voxel size of the transform."""
+        return tuple(x * y for x, y in zip(self.scale, self.units))
```
Comment (jni) [inline napari/utils/transforms/transforms.py]: I don't like voxel_size because it's not semantically very different from "scale" — it doesn't immediately tell me that it's got units attached. Also, I must admit I was confused by the use of "x * y" (the variable names sound like axis names) in this context. 😅 So I propose:

```suggestion
    @property
    def physical_scale(self) -> tuple[pint.Quantity, ...]:
        """Return the scale of the transform, with units."""
        return tuple(magnitude * unit for magnitude, unit in zip(self.scale, self.units))
```

Comment (jni) [reply napari/utils/transforms/transforms.py]: Alternative:

```suggestion
    @property
    def physical_scale(self) -> tuple[pint.Quantity, ...]:
        """Return the scale of the transform, with units."""
        return tuple(np.multiply(self.scale, self.units))
```

File: napari/utils/transforms/_tests/test_transforms.py (hunk 1 above)
Comment (Czaki) [reply napari/utils/transforms/_tests/test_transforms.py:14]: I do not know. But it is builtin. 
https://github.com/hgrecco/pint/blob/master/pint/default_en.txt#L671

**Comment by jni** [comment]
@Czaki thanks for picking this up! I really like this minimal PR, I think it's very close to ready-to-merge!

It's a bit annoying to see `pint.get_application_registry()` everywhere — *maybe* we should have `napari.transforms.units` as an alias for that registry? (Yes I am suggesting a top-level transforms module — I think it will be useful enough to have.) But this is a minor issue and maybe we can postpone this code style decision to a later PR.

**Comment by Czaki** [comment]
> @Czaki thanks for picking this up! I really like this minimal PR, I think it's very close to ready-to-merge!

I need to add more tests 

> It's a bit annoying to see `pint.get_application_registry()` everywhere — _maybe_ we should have `napari.transforms.units` as an alias for that registry? (Yes I am suggesting a top-level transforms module — I think it will be useful enough to have.) But this is a minor issue and maybe we can postpone this code style decision to a later PR.

pint allows replacing application register and does not allow mixing units from different registers. So I prefer to write code that allows for such change. 
I have hope that plugins will not do this, but someone who would like to embed napari as viewer in an application may want to use this option. So I prefer not to determine register on import time.

**Comment by jni** [comment]
> pint allows replacing application register and does not allow mixing units from different registers. So I prefer to write code that allows for such change.
> I have hope that plugins will not do this, but someone who would like to embed napari as viewer in an application may want to use this option. So I prefer not to determine register on import time.

Ok so we just need to write a PEP that allows module level properties (similar to module getattr) and wait a few years and then we're good! 😜

**Comment by Czaki** [comment]
propery will have same problem as `getattr`, swo when someone will use `from napari.transforms.units import pixel` will be resolved on import time. To solve this, the `napari.transforms.units` need to be an object, not module.

File: napari/utils/transforms/_tests/test_transforms.py (hunk 2 above)
Comment (Czaki) [reply napari/utils/transforms/_tests/test_transforms.py]: updated

File: napari/utils/transforms/_tests/test_transforms.py (hunk 1 above)
Comment (jni) [reply napari/utils/transforms/_tests/test_transforms.py:14]: From git-blame it seems ancient (8+ years). I must have missed it somehow. 😅 Or I'm thinking of a different unit... Anyway, cool!

**Comment by jni** [comment]
Please ping me when this is ready for review @Czaki 😅

**Comment by Czaki** [comment]
> Please ping me when this is ready for review @Czaki 😅

Ready to merge was added to test Rest API for 6984

**Comment by Czaki** [comment]
I do not understand windows fail. I will need to debug this, but review may be done now as it should be a pure technical thing.

**Review by brisvag — COMMENTED** [review]
I like where this is going!

Was there any discussion related to and accounting for https://github.com/napari/napari/issues/3783?

**Comment by Czaki** [comment]
> Was there any discussion related to and accounting for #3783?

No. There was a request to do minimal PR to expose a new API and allow use it in 0.5.0, then improve it in followup PR that may land in 0.5.x

**Comment by brisvag** [comment]
Ok, just wanted to make sure there was an eye on it :)

**Comment by jni** [comment]
@brisvag we are missing you at the meetings! 😃 

Yeah the goal here is to take some baby steps and allow plugins to start putting in axis and units information in the LayerDataTuple. Later, when we work on a proper transformations architecture, we can give plugins the right deprecation messages. But in the meantime we can start using the information and we get the community unstuck with providing us with this info.

@Czaki I'll review this either later today or tomorrow. Thanks!!! 🙏

File: napari/layers/_layer_actions.py [hunk 6]
```diff
@@ -191,7 +191,16 @@ def _project(ll: LayerList, axis: int = 0, mode: str = 'max') -> None:
     meta = {
         key: layer._get_base_state()[key]
         for key in layer._get_base_state()
-        if key not in ('scale', 'translate', 'rotate', 'shear', 'affine')
```
Comment (jni) [inline napari/layers/_layer_actions.py:194]: This is actually incorrect, but not the fault of this PR... Should we make an issue? Projection should *at least* subsample the transforms for the given axes. (It's more complicated if there are transformations that are not orthogonal to the projection axis, but that can tackle those separately.)

File: napari/layers/base/_tests/test_base.py [hunk 7]
```diff
@@ -0,0 +1,106 @@
...
+    assert layer.units == (REG.pixel, REG.pixel)
+
+
+def test_assign_units_error():
+    layer = SampleLayer(np.empty((10, 10)))
+    with pytest.raises(ValueError, match='must have length ndim'):
+        layer.units = ('m', 'm', 'm')
+
+    with pytest.raises(ValueError, match='Could not find unit'):
+        layer.units = ('ugh', 'ugh')
```
Comment (jni) [inline napari/layers/base/_tests/test_base.py:63]: 😂

File: napari/layers/base/base.py [hunk 8]
```diff
@@ -748,6 +764,32 @@ def _reset_editable(self) -> None:
...
+        """List of units for the layer."""
+        return self._transforms['data2physical'].units
+
+    @units.setter
+    def units(self, units: Optional[Sequence[pint.Unit]]) -> None:
+        prev = self.units
+        # mypy bug https://github.com/python/mypy/issues/3004
+        self._transforms['data2physical'].units = units  # type: ignore[assignment]
+        if self.units != prev:
+            self.events.units()
```
Comment (jni) [inline napari/layers/base/base.py:791]: I'm never really sure about whether events should include their value — I think yes? Should this be `self.events.units(value=self.units)`? (Happy to have this discussion in a follow-up issue/PR.)

File: napari/layers/utils/stack_utils.py [hunk 9]
```diff
@@ -287,6 +292,8 @@ def images_to_stack(images: list[Image], axis: int = 0, **kwargs) -> Image:
     kwargs.setdefault('translate', np.insert(meta['translate'], axis, 0))
 
     meta.update(kwargs)
+    meta['units'] = (pint.get_application_registry().pixel,) + meta['units']
```
Comment (jni) [inline napari/layers/utils/stack_utils.py:295]: 👍 clever. I like this. 😊

File: napari/utils/transforms/_units.py [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+@overload
+def get_units_from_name(
+    units: Sequence[Union[str, pint.Unit]],
+) -> tuple[pint.Unit, ...]: ...
+
+
+def get_units_from_name(units: UnitsLike) -> UnitsInfo:
+    """
+    Convert a string or sequence of strings to pint units.
+    """
```
Comment (jni) [inline napari/utils/transforms/_units.py]:
```suggestion
    """Convert a string or sequence of strings to pint units."""
```

Comment (jni) [reply napari/utils/transforms/_units.py]: [PEP-257 on one-liners](https://peps.python.org/pep-0257/#one-line-docstrings):

> The closing quotes are on the same line as the opening quotes.

File: napari/utils/transforms/transforms.py [hunk 11]
```diff
@@ -491,6 +500,38 @@ def ndim(self) -> int:
...
 
+    @property
+    def axis_labels(self) -> tuple[str, ...]:
+        """tuple of axis labels for the layer."""
+        return self._axis_labels
+
+    @axis_labels.setter
+    def axis_labels(self, axis_labels: Optional[Sequence[str]]) -> None:
+        if axis_labels is None:
+            axis_labels = tuple(f'axis {i}' for i in range(-self.ndim, 0))
```
Comment (jni) [inline napari/utils/transforms/transforms.py:511]: I'm not sure here, just a proposal for discussion:
1. `'axis -i'` is one option (current option)
2. *Just* `'-i'`
3. directly integers instead of strings, so they can be used for indexing.

I'm not sure at all of the merits of each option — I'm proposing only because in my brain I probably would have used option (3). I suspect that it does not much matter because we will never be able to actually to use them directly — we'll need to write code that does different things depending on the input. However, for example, it might be interesting to use just the integers, because then you know whether you are using the default or something set by the metadata. The same can be said of option 2, if we for example forbid strings starting with `-` from users.

Thoughts?

File: tox.ini [hunk 12]
```diff
@@ -103,7 +103,7 @@ commands =
     echo "COVERAGE: {env:COVERAGE:}"
     cov: coverage run \
     !cov: python \
-        -m pytest -v {env:PYTEST_PATH:} --color=yes --basetemp={envtmpdir} \
+        -m pytest {env:PYTEST_PATH:} --color=yes --basetemp={envtmpdir} \
```
Comment (jni) [inline tox.ini]: Seems unrelated... Why?

**Review by jni — APPROVED** [review]
@Czaki nice! I've made a few minor comments + points for discussion, but I'd be happy if this went in as-is, so I'm going to mark as ready-to-merge. 🚀

File: napari/layers/base/base.py (hunk 8 above)
Comment (Czaki) [reply napari/layers/base/base.py:791]: We need to first resolve the problem that we discussed here: https://github.com/pyapp-kit/psygnal/pull/296#pullrequestreview-1948842934
Otherwise, signal could contain wrong value of units (different than current one). 

It may be a part of migration to psygnal.

File: napari/utils/transforms/transforms.py (hunk 11 above)
Comment (Czaki) [reply napari/utils/transforms/transforms.py:511]: I do not want ints. I prefer to keep it sting only. Code will be simpler and easier to avoid errors.

File: tox.ini (hunk 12 above)
Comment (Czaki) [reply tox.ini]: Because with almost 4k tests, the logs are unreadable. Finding test that fails is harder than in non verbose mode.

File: napari/layers/_layer_actions.py (hunk 6 above)
Comment (Czaki) [reply napari/layers/_layer_actions.py:194]: Projection subsamples transforms in lines 214-216 of this file. Should I add a comment with information about this (I need a few minutest to understand this when writing code).

File: tox.ini (hunk 12 above)
Comment (jni) [reply tox.ini]: Can we make the change in a separate PR?

Comment (Czaki) [reply tox.ini]: done #6996

Comment (jni) [reply tox.ini]: #6996 Thanks!

File: napari/layers/_layer_actions.py (hunk 6 above)
Comment (jni) [reply napari/layers/_layer_actions.py:194]: I think a comment would be very useful! 👍

File: napari/utils/transforms/transforms.py (hunk 11 above)
Comment (jni) [reply napari/utils/transforms/transforms.py:511]: Ok. How about option 2 then?

Comment (Czaki) [reply napari/utils/transforms/transforms.py:511]: I prefer `axis -1` as it is less possible to make confusion when printing. But do not have a strong opinion.

Comment (jni) [reply napari/utils/transforms/transforms.py:511]: I just think that at some point these are going to end up on the dims controls on the viewer, and "axis -1" takes up so much space/is ugly. I agree that it's less ambiguous, though.

Since currently we are not using this information anywhere, I am going to merge this and make an issue about it. (After merging in main to get CI passing again.)

**Comment by jni** [comment]
Hmmm that windows error is pretty stubborn @Czaki 😬

**Comment by jni** [comment]
[This](https://github.com/napari/napari/actions/runs/9546502720/job/26310351763?pr=6979#step:12:461) is the error:

```
  ================================== FAILURES ===================================
  __________________________ test_all_supported_dtypes __________________________
  
  qt_viewer = <napari._qt.qt_viewer.QtViewer object at 0x0000022E76799240>
  
      def test_all_supported_dtypes(qt_viewer):
          data = np.zeros((10, 10), dtype=np.uint8)
          layer = qt_viewer.viewer.add_labels(data, opacity=1)
      
          for i, dtype in enumerate(NUMPY_INTEGER_TYPES, start=1):
              data = np.full((10, 10), i, dtype=dtype)
              layer.data = data
              QApplication.processEvents()
              canvas_screenshot = qt_viewer.screenshot(flash=False)
              midd_pixel = canvas_screenshot[
                  tuple(np.array(canvas_screenshot.shape[:2]) // 2)
              ]
  >           npt.assert_equal(
                  midd_pixel, layer.colormap.map(i) * 255, err_msg=f'{dtype} {i}'
              )
  
  napari\_qt\_tests\test_qt_viewer.py:1017: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  
  args = (<built-in function eq>, array([  0,   0,   0, 255], dtype=uint8), ColorArray([108.,   2., 193., 255.], dtype=float32))
  kwds = {'err_msg': "<class 'numpy.int64'> 4", 'header': 'Arrays are not equal', 'strict': False, 'verbose': True}
  
      @wraps(func)
      def inner(*args, **kwds):
          with self._recreate_cm():
  >           return func(*args, **kwds)
  E           AssertionError: 
  E           Arrays are not equal
  E           <class 'numpy.int64'> 4
  E           Mismatched elements: 3 / 4 (75%)
  E           Max absolute difference: 193.
  E           Max relative difference: 1.
  E            x: array([  0,   0,   0, 255], dtype=uint8)
  E            y: ColorArray([108.,   2., 193., 255.], dtype=float32)
  
  C:\hostedtoolcache\windows\Python\3.10.11\x64\lib\contextlib.py:79: AssertionError
```

It's a weird one... Like events are not being processed correctly? 🤔

**Comment by Czaki** [comment]
> Hmmm that windows error is pretty stubborn @Czaki 😬

Yes. Will dig it today. Remove `ready to merge` for now. 

> It's a weird one... Like events are not being processed correctly? 🤔

Yes. It looks like the canvas is not repainted.

**Comment by jni** [comment]
I'm not super opposed to skipping/xfailing the test in windows CI, if you can check manually that it works in a Windows machine...

**Comment by jni** [comment]
It's *so weird* that this PR would trigger this error. Do you think it's some fragility in the tests that we're exposing with new tests? Or did an earlier PR cause this error and we missed it? 🤔

File: napari/layers/_layer_actions.py (hunk 6 above)
Comment (Czaki) [reply napari/layers/_layer_actions.py:194]: is it ok now?

**Comment by Czaki** [comment]
Error log also contains this:

```
 Traceback (most recent call last):
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\app\backends\_qt.py", line 928, in paintGL
      self._vispy_canvas.events.draw(region=None)
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\util\event.py", line 453, in __call__
      self._invoke_callback(cb, event)
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\util\event.py", line 471, in _invoke_callback
      _handle_exception(self.ignore_callback_errors,
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\util\event.py", line 469, in _invoke_callback
      cb(event)
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\scene\canvas.py", line 219, in on_draw
      self._draw_scene()
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\scene\canvas.py", line 278, in _draw_scene
      self.draw_visual(self.scene)
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\scene\canvas.py", line 316, in draw_visual
      node.draw()
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\scene\visuals.py", line 106, in draw
      self._visual_superclass.draw(self)
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\visuals\visual.py", line 514, in draw
      self._program.draw(self._vshare.draw_mode,
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\visuals\shaders\program.py", line 102, in draw
      Program.draw(self, *args, **kwargs)
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\gloo\program.py", line 543, in draw
      canvas.context.flush_commands()
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\gloo\context.py", line 172, in flush_commands
      self.glir.flush(self.shared.parser)
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\gloo\glir.py", line 584, in flush
      self._shared.flush(parser)
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\gloo\glir.py", line 506, in flush
      parser.parse(self._filter(self.clear(), parser))
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\gloo\glir.py", line 824, in parse
      self._parse(command)
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\gloo\glir.py", line 794, in _parse
      ob.set_data(*args)
    File "D:\a\napari\napari\.tox\py310-windows-pyqt5-no_cov\lib\site-packages\vispy\gloo\glir.py", line 933, in set_data
      raise RuntimeError("Shader compilation error in %s:\n%s" %
  RuntimeError: Shader compilation error in GL_VERTEX_SHADER:
```

File: napari/utils/transforms/transforms.py (hunk 11 above)
Comment (Czaki) [reply napari/utils/transforms/transforms.py:511]: I think that in some moment, we will land in ordering `t`, `z`, `x`, `y` based on units metadata with addition of `p`, `m`, `s`  (or other letters used in tiff metadata)

**Comment by brisvag** [comment]
Not sure if it can be related, but the last run uses the new vispy release, while the previous (failing) one pulled the previous vispy... :shrug:

**Comment by jni** [comment]
@Czaki I turned off the requirement to be up to date. Will look into merge queues tomorrow as an alternative.

**Comment by jni** [comment]
(I must say now that we have constraints I have seen much fewer sporadic failures! 🥳 Should make merge queues much easier to use.)

**Comment by jni** [comment]
(Feel free to add ready-to-merge again if the tests pass!)

**Comment by Czaki** [comment]
I do not understand why this change solves this problem, but will not touch it for now.
